ambry-store/src/test/java/com.github.ambry.store/BlobStoreStatsTest.java

```diff
@@ -241,8 +315,351 @@ public void testValidDataSizeAfterDeletes() throws InterruptedException, StoreEx
+    secondCheckpointHoldLatch.countDown();
+    // a probe put with a latch to inform us about the state of the queue
+    CountDownLatch queueProcessedLatch = new CountDownLatch(1);
+    blobStoreStats.handleNewPutEntry(new MockIndexValue(queueProcessedLatch, state.index.getCurrentEndOffset()));
+    assertTrue("QueueProcessor took too long to process the new entries",
+        queueProcessedLatch.await(5, TimeUnit.SECONDS));
+    for (long i = 0; i <= state.time.milliseconds() + TEST_TIME_INTERVAL_IN_MS; i += TEST_TIME_INTERVAL_IN_MS) {
+      verifyAndGetLogSegmentValidSize(blobStoreStats, new TimeRange(i, 0L));
+    }
+    verifyAndGetContainerValidSize(blobStoreStats, state.time.milliseconds());
+    long timeToLiveInMs = expiresAtInMs - state.time.milliseconds() < 0 ? 0 : expiresAtInMs - state.time.milliseconds();
+    state.advanceTime(timeToLiveInMs + Time.MsPerSec);
+    advanceTimeToNextSecond();
+    verifyAndGetContainerValidSize(blobStoreStats, state.time.milliseconds());
+    blobStoreStats.close();
+  }
+
+  /**
+   * Tests to verify requests inside and outside of the forecast coverage can still be served properly while scanning.
+   * @throws InterruptedException
+   * @throws StoreException
+   * @throws IOException
+   */
+  @Test
+  public void testBucketingCoverageTransition() throws InterruptedException, StoreException, IOException {
+    assumeTrue(bucketingEnabled);
+    CountDownLatch scanStartedLatch = new CountDownLatch(1);
+    // do not hold the initial scan
+    MockThrottler mockThrottler = new MockThrottler(scanStartedLatch, new CountDownLatch(0));
+    throttlers.put(BlobStoreStats.IO_SCHEDULER_JOB_TYPE, mockThrottler);
+    // add a put that is going to expire in 20 seconds
+    long initialScanTimeInMs = state.time.milliseconds();
+    BlobStoreStats blobStoreStats = setupBlobStoreStats(2, 0);
+    // proceed only when the scan is started
+    assertTrue("IndexScanner took too long to start", scanStartedLatch.await(5, TimeUnit.SECONDS));
+    verifyAndGetLogSegmentValidSize(blobStoreStats, new TimeRange(initialScanTimeInMs, 0L));
+    verifyAndGetContainerValidSize(blobStoreStats, initialScanTimeInMs);
+    // advance time by 20 seconds
+    state.advanceTime(TimeUnit.SECONDS.toMillis(20));
+    // hold the next scan
+    CountDownLatch scanHoldLatch = new CountDownLatch(1);
+    scanStartedLatch = new CountDownLatch(1);
+    mockThrottler.holdLatch = scanHoldLatch;
+    mockThrottler.startedLatch = scanStartedLatch;
+    mockThrottler.isThrottlerStarted = false;
+    assertTrue("IndexScanner took too long to start", scanStartedLatch.await(5, TimeUnit.SECONDS));
+    int throttleCountBeforeRequests = mockThrottler.throttleCount.get();
+    verifyAndGetLogSegmentValidSize(blobStoreStats, new TimeRange(initialScanTimeInMs, 0L));
+    verifyAndGetContainerValidSize(blobStoreStats, initialScanTimeInMs);
+    assertEquals("Throttle count shouldn't have changed", throttleCountBeforeRequests,
```

Comment:
how can you be sure that the index scanning is complete. scanStartedLatch tell us whether the scanning has started or not. It doesn't say whether the scanning is complete right.